File: javascript/src/runtime-checkpoint-v2.mjs
import { canonicalHash, canonicalJson } from "./canonical.mjs";
import { parseStrictJson } from "./strict-json.mjs";
import { ScriptRuntimeV3 } from "./runtime-v3.mjs";
import { decodeV3Value, encodeV3Value } from "./ir-v3-values.mjs";
import { validateProgramIrV3 } from "./ir-v3-validation.mjs";
import { TevScriptError } from "./values.mjs";

const SCHEMA = "TEV_SCRIPT_RUNTIME_CHECKPOINT_V2";
const LOCAL = /^[A-Za-z_][A-Za-z0-9_]*$/;
const SHA = /^[0-9a-f]{64}$/;
const ROOT_KEYS = Object.freeze([
  "entities", "ir_schema", "program_id", "schema", "semantic_hash",
  "source_schema", "source_semantic_hash",
]);

function fail(code, path, message) {
  throw new TevScriptError(code, `${path}: ${message}`);
}
function requireObject(value, path) {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    fail("TEVS_CHECKPOINT_V2_SHAPE", path, "expected object");
  }
  return value;
}
function exactKeys(value, path, expected) {
  const observed = Object.keys(value).sort();
  const wanted = [...expected].sort();
  if (observed.length !== wanted.length || observed.some((item, index) => item !== wanted[index])) {
    fail("TEVS_CHECKPOINT_V2_SHAPE", path, `field set mismatch; expected=${wanted.join(",")}, observed=${observed.join(",")}`);
  }
}
function requireString(value, path) {
  if (typeof value !== "string") fail("TEVS_CHECKPOINT_V2_SHAPE", path, "expected string");
  return value;
}
function requireLocal(value, path) {
  const result = requireString(value, path);
  if (!LOCAL.test(result)) fail("TEVS_CHECKPOINT_V2_IDENTIFIER", path, `expected identifier, got ${JSON.stringify(result)}`);
  return result;
}
function requireSha(value, path) {
  const result = requireString(value, path);
  if (!SHA.test(result)) fail("TEVS_CHECKPOINT_V2_SHA256", path, "expected lowercase SHA-256");
  return result;
}

export class RuntimeCheckpointV2 {
  constructor({ programId, irSchema, semanticHash, sourceSchema, sourceSemanticHash, entities }) {
    this.programId = programId;
    this.irSchema = irSchema;
    this.semanticHash = semanticHash;
    this.sourceSchema = sourceSchema;
    this.sourceSemanticHash = sourceSemanticHash;
    this.entities = entities.map(([entityId, state]) => [
      entityId,
      state.map(([name, typeId, value]) => [name, typeId, structuredClone(value)]),
    ]);
    Object.freeze(this.entities);
    Object.freeze(this);
  }

  static capture(runtime) {
    const entities = [];
    for (const entityId of [...runtime.entities.keys()].sort()) {
      const entity = runtime.entities.get(entityId);
      const state = [];
      for (const stateName of [...entity.state.keys()].sort()) {
        const typeId = entity.stateTypes.get(stateName);
        state.push([
          stateName,
          typeId,
          encodeV3Value(typeId, entity.state.get(stateName), runtime.typeTable, {
            context: `checkpoint ${entityId}.${stateName}`,
          }),
        ]);
      }
      entities.push([entityId, state]);
    }
    return new RuntimeCheckpointV2({
      programId: runtime.ir.program_id,
      irSchema: runtime.ir.schema,
      semanticHash: runtime.ir.semantic_hash,
      sourceSchema: runtime.ir.source_schema,
      sourceSemanticHash: runtime.ir.source_semantic_hash,
      entities,
    });
  }

  static parse(text) {
    const parsed = parseStrictJson(text);
    if (canonicalJson(parsed) !== text) {
      throw new TevScriptError(
        "TEVS_CHECKPOINT_V2_CANONICAL",
        "runtime checkpoint must use exact canonical JSON bytes",
      );
    }
    const root = requireObject(parsed, "$");
    exactKeys(root, "$", ROOT_KEYS);
    if (root.schema !== SCHEMA) fail("TEVS_CHECKPOINT_V2_SCHEMA", "$.schema", `expected ${SCHEMA}`);
    const programId = requireLocal(root.program_id, "$.program_id");
    const irSchema = requireString(root.ir_schema, "$.ir_schema");
    if (irSchema !== "TEV_SCRIPT_PROGRAM_IR_V3") {
      fail("TEVS_CHECKPOINT_V2_IR_SCHEMA", "$.ir_schema", "expected TEV_SCRIPT_PROGRAM_IR_V3");
    }
    const semanticHash = requireSha(root.semantic_hash, "$.semantic_hash");
    const sourceSchema = requireString(root.source_schema, "$.source_schema");
    if (sourceSchema !== "TEV_SCRIPT_LINKED_PROGRAM_V1" && sourceSchema !== "TEV_SCRIPT_PROGRAM_IR_V2") {
      fail("TEVS_CHECKPOINT_V2_SOURCE_SCHEMA", "$.source_schema", `unsupported source schema ${JSON.stringify(sourceSchema)}`);
    }
    const sourceSemanticHash = requireSha(root.source_semantic_hash, "$.source_semantic_hash");
    if (!Array.isArray(root.entities)) fail("TEVS_CHECKPOINT_V2_SHAPE", "$.entities", "expected array");
    const entities = [];
    let previous = null;
    const seen = new Set();
    root.entities.forEach((rawEntity, index) => {
      const path = `$.entities[${index}]`;
      const entity = requireObject(rawEntity, path);
      exactKeys(entity, path, ["entity_id", "state"]);
      const entityId = requireLocal(entity.entity_id, `${path}.entity_id`);
      if (seen.has(entityId) || (previous !== null && entityId <= previous)) {
        fail("TEVS_CHECKPOINT_V2_ENTITY_ORDER", `${path}.entity_id`, "entities must be unique and strictly sorted");
      }
      seen.add(entityId);
      previous = entityId;
      const rawState = requireObject(entity.state, `${path}.state`);
      const state = [];
      for (const stateName of Object.keys(rawState).sort()) {
        requireLocal(stateName, `${path}.state key`);
        const typed = requireObject(rawState[stateName], `${path}.state.${stateName}`);
        exactKeys(typed, `${path}.state.${stateName}`, ["type", "value"]);
        state.push([
          stateName,
          requireString(typed.type, `${path}.state.${stateName}.type`),
          structuredClone(typed.value),
        ]);
      }
      entities.push([entityId, state]);
    });
    return new RuntimeCheckpointV2({
      programId, irSchema, semanticHash, sourceSchema, sourceSemanticHash, entities,
    });
  }

  toObject() {
    return {
      schema: SCHEMA,
      program_id: this.programId,
      ir_schema: this.irSchema,
      semantic_hash: this.semanticHash,
      source_schema: this.sourceSchema,
      source_semantic_hash: this.sourceSemanticHash,
      entities: this.entities.map(([entityId, state]) => ({
        entity_id: entityId,
        state: Object.fromEntries(state.map(([name, typeId, value]) => [
          name,
          { type: typeId, value: structuredClone(value) },
        ])),
      })),
    };
  }

  toCanonicalJson() { return canonicalJson(this.toObject()); }
  get checkpointHash() { return canonicalHash(this.toObject()); }

  restoreExact(ir, capabilities = {}) {
    const target = structuredClone(ir);
    const table = validateProgramIrV3(target, {
      expectedSourceSemanticHash: this.sourceSemanticHash,
    });
    if (target.program_id !== this.programId) fail("TEVS_CHECKPOINT_V2_PROGRAM_ID", "$.program_id", "checkpoint program id does not match target");
    if (target.schema !== this.irSchema) fail("TEVS_CHECKPOINT_V2_IR_SCHEMA", "$.ir_schema", "checkpoint IR schema does not match target");
    if (target.semantic_hash !== this.semanticHash) fail("TEVS_CHECKPOINT_V2_SEMANTIC_HASH", "$.semantic_hash", "checkpoint semantic hash does not match target");
    if (target.source_schema !== this.sourceSchema) fail("TEVS_CHECKPOINT_V2_SOURCE_SCHEMA", "$.source_schema", "checkpoint source schema does not match target");
    if (target.source_semantic_hash !== this.sourceSemanticHash) fail("TEVS_CHECKPOINT_V2_SOURCE_HASH", "$.source_semantic_hash", "checkpoint source hash does not match target");

    const expectedEntities = new Map(target.entities.map((entity) => [entity.entity_id, entity]));
    const checkpointEntities = new Map(this.entities);
    if (expectedEntities.size !== checkpointEntities.size
        || [...expectedEntities.keys()].some((id) => !checkpointEntities.has(id))) {
      fail("TEVS_CHECKPOINT_V2_ENTITY_SET", "$.entities", "checkpoint entity set does not exactly match target");
    }

    const decodedState = new Map();
    for (const entityId of [...expectedEntities.keys()].sort()) {
      const definition = expectedEntities.get(entityId);
      const expectedStates = new Map(definition.states.map((state) => [state.name, state.type]));
      const checkpointState = new Map(checkpointEntities.get(entityId).map(([name, typeId, value]) => [name, [typeId, value]]));
      if (expectedStates.size !== checkpointState.size
          || [...expectedStates.keys()].some((name) => !checkpointState.has(name))) {
        fail("TEVS_CHECKPOINT_V2_STATE_SET", `entity ${entityId}`, "checkpoint state set does not exactly match target");
      }
      const values = new Map();
      for (const stateName of [...expectedStates.keys()].sort()) {
        const expectedType = expectedStates.get(stateName);
        const [observedType, rawValue] = checkpointState.get(stateName);
        if (observedType !== expectedType) {
          fail("TEVS_CHECKPOINT_V2_STATE_TYPE", `${entityId}.${stateName}`, `expected ${expectedType}, got ${observedType}`);
        }
        values.set(
          stateName,
          decodeV3Value(expectedType, rawValue, table, { context: `checkpoint ${entityId}.${stateName}` }),
        );
      }
      decodedState.set(entityId, values);
    }

    const runtime = new ScriptRuntimeV3(target, capabilities, {
      expectedSourceSemanticHash: this.sourceSemanticHash,
    });
    for (const [entityId, values] of decodedState) {
      const entity = runtime.entities.get(entityId);
      entity.state.clear();
      for (const [name, value] of values) entity.state.set(name, value);
    }
    return runtime;
  }
}

export const RUNTIME_CHECKPOINT_V2_SCHEMA = SCHEMA;

File: tev_script/runtime_checkpoint_v2.py
from __future__ import annotations

from dataclasses import dataclass
import re
from typing import Any, Mapping

from .canonical import canonical_hash, canonical_json
from .diagnostics import TevScriptError
from .ir_v3_validation import validate_program_ir_v3
from .ir_v3_values import decode_v3_value, encode_v3_value
from .json_io import parse_strict_json
from .runtime_v3 import CapabilityV3, ScriptRuntimeV3

_SCHEMA = "TEV_SCRIPT_RUNTIME_CHECKPOINT_V2"
_HASH = re.compile(r"^[0-9a-f]{64}$")
_LOCAL = re.compile(r"^[A-Za-z_][A-Za-z0-9_]*$")
_ROOT_KEYS = {
    "schema", "program_id", "ir_schema", "semantic_hash",
    "source_schema", "source_semantic_hash", "entities",
}


@dataclass(frozen=True, slots=True)
class RuntimeCheckpointV2:
    program_id: str
    ir_schema: str
    semantic_hash: str
    source_schema: str
    source_semantic_hash: str
    entities: tuple[tuple[str, tuple[tuple[str, str, Any], ...]], ...]

    @classmethod
    def capture(cls, runtime: ScriptRuntimeV3) -> "RuntimeCheckpointV2":
        entities: list[tuple[str, tuple[tuple[str, str, Any], ...]]] = []
        for entity_id in sorted(runtime.entities):
            entity = runtime.entities[entity_id]
            state: list[tuple[str, str, Any]] = []
            for state_name in sorted(entity.state):
                type_id = entity.state_types[state_name]
                state.append(
                    (
                        state_name,
                        type_id,
                        encode_v3_value(
                            type_id,
                            entity.state[state_name],
                            runtime.type_table,
                            context=f"checkpoint {entity_id}.{state_name}",
                        ),
                    )
                )
            entities.append((entity_id, tuple(state)))
        return cls(
            str(runtime.ir["program_id"]),
            str(runtime.ir["schema"]),
            str(runtime.ir["semantic_hash"]),
            str(runtime.ir["source_schema"]),
            str(runtime.ir["source_semantic_hash"]),
            tuple(entities),
        )

    @classmethod
    def parse(cls, text: str) -> "RuntimeCheckpointV2":
        parsed = parse_strict_json(text)
        canonical = canonical_json(parsed)
        if canonical != text:
            raise TevScriptError(
                "TEVS_CHECKPOINT_V2_CANONICAL",
                "runtime checkpoint must use exact canonical JSON bytes",
            )
        root = _object(parsed, "$")
        _exact_keys(root, "$", _ROOT_KEYS)
        if root["schema"] != _SCHEMA:
            _fail("TEVS_CHECKPOINT_V2_SCHEMA", "$.schema", f"expected {_SCHEMA}")
        program_id = _local(root["program_id"], "$.program_id")
        ir_schema = _string(root["ir_schema"], "$.ir_schema")
        if ir_schema != "TEV_SCRIPT_PROGRAM_IR_V3":
            _fail("TEVS_CHECKPOINT_V2_IR_SCHEMA", "$.ir_schema", "expected TEV_SCRIPT_PROGRAM_IR_V3")
        semantic_hash = _sha(root["semantic_hash"], "$.semantic_hash")
        source_schema = _string(root["source_schema"], "$.source_schema")
        if source_schema not in {"TEV_SCRIPT_LINKED_PROGRAM_V1", "TEV_SCRIPT_PROGRAM_IR_V2"}:
            _fail("TEVS_CHECKPOINT_V2_SOURCE_SCHEMA", "$.source_schema", f"unsupported source schema {source_schema!r}")
        source_hash = _sha(root["source_semantic_hash"], "$.source_semantic_hash")

        raw_entities = _array(root["entities"], "$.entities")
        entities: list[tuple[str, tuple[tuple[str, str, Any], ...]]] = []
        previous_entity: str | None = None
        entity_ids: set[str] = set()
        for index, raw_entity in enumerate(raw_entities):
            path = f"$.entities[{index}]"
            entity = _object(raw_entity, path)
            _exact_keys(entity, path, {"entity_id", "state"})
            entity_id = _local(entity["entity_id"], path + ".entity_id")
            if entity_id in entity_ids or (previous_entity is not None and entity_id <= previous_entity):
                _fail("TEVS_CHECKPOINT_V2_ENTITY_ORDER", path + ".entity_id", "entities must be unique and strictly sorted")
            entity_ids.add(entity_id)
            previous_entity = entity_id
            raw_state = _object(entity["state"], path + ".state")
            state: list[tuple[str, str, Any]] = []
            for state_name in sorted(raw_state):
                _local(state_name, path + ".state key")
                typed = _object(raw_state[state_name], path + f".state.{state_name}")
                _exact_keys(typed, path + f".state.{state_name}", {"type", "value"})
                state.append((state_name, _string(typed["type"], path + f".state.{state_name}.type"), typed["value"]))
            entities.append((entity_id, tuple(state)))
        return cls(program_id, ir_schema, semantic_hash, source_schema, source_hash, tuple(entities))

    def to_object(self) -> dict[str, object]:
        return {
            "schema": _SCHEMA,
            "program_id": self.program_id,
            "ir_schema": self.ir_schema,
            "semantic_hash": self.semantic_hash,
            "source_schema": self.source_schema,
            "source_semantic_hash": self.source_semantic_hash,
            "entities": [
                {
                    "entity_id": entity_id,
                    "state": {
                        state_name: {"type": type_id, "value": value}
                        for state_name, type_id, value in state
                    },
                }
                for entity_id, state in self.entities
            ],
        }

    def to_canonical_json(self) -> str:
        return canonical_json(self.to_object())

    @property
    def checkpoint_hash(self) -> str:
        return canonical_hash(self.to_object())

    def restore_exact(
        self,
        ir: Mapping[str, Any],
        capabilities: Mapping[str, CapabilityV3] | None = None,
    ) -> ScriptRuntimeV3:
        target = dict(ir)
        table = validate_program_ir_v3(
            target,
            expected_source_semantic_hash=self.source_semantic_hash,
        )
        if str(target["program_id"]) != self.program_id:
            _fail("TEVS_CHECKPOINT_V2_PROGRAM_ID", "$.program_id", "checkpoint program id does not match target")
        if str(target["schema"]) != self.ir_schema:
            _fail("TEVS_CHECKPOINT_V2_IR_SCHEMA", "$.ir_schema", "checkpoint IR schema does not match target")
        if str(target["semantic_hash"]) != self.semantic_hash:
            _fail("TEVS_CHECKPOINT_V2_SEMANTIC_HASH", "$.semantic_hash", "checkpoint semantic hash does not match target")
        if str(target["source_schema"]) != self.source_schema:
            _fail("TEVS_CHECKPOINT_V2_SOURCE_SCHEMA", "$.source_schema", "checkpoint source schema does not match target")
        if str(target["source_semantic_hash"]) != self.source_semantic_hash:
            _fail("TEVS_CHECKPOINT_V2_SOURCE_HASH", "$.source_semantic_hash", "checkpoint source hash does not match target")

        expected_entities = {str(entity["entity_id"]): entity for entity in target["entities"]}
        checkpoint_entities = {entity_id: state for entity_id, state in self.entities}
        if set(expected_entities) != set(checkpoint_entities):
            _fail("TEVS_CHECKPOINT_V2_ENTITY_SET", "$.entities", "checkpoint entity set does not exactly match target")

        decoded_state: dict[str, dict[str, Any]] = {}
        for entity_id in sorted(expected_entities):
            definition = expected_entities[entity_id]
            expected_states = {str(state["name"]): str(state["type"]) for state in definition["states"]}
            checkpoint_state = {name: (type_id, value) for name, type_id, value in checkpoint_entities[entity_id]}
            if set(expected_states) != set(checkpoint_state):
                _fail("TEVS_CHECKPOINT_V2_STATE_SET", f"entity {entity_id}", "checkpoint state set does not exactly match target")
            decoded_state[entity_id] = {}
            for state_name in sorted(expected_states):
                expected_type = expected_states[state_name]
                observed_type, raw_value = checkpoint_state[state_name]
                if observed_type != expected_type:
                    _fail(
                        "TEVS_CHECKPOINT_V2_STATE_TYPE",
                        f"{entity_id}.{state_name}",
                        f"expected {expected_type}, got {observed_type}",
                    )
                decoded_state[entity_id][state_name] = decode_v3_value(
                    expected_type,
                    raw_value,
                    table,
                    context=f"checkpoint {entity_id}.{state_name}",
                )

        runtime = ScriptRuntimeV3(
            target,
            capabilities,
            expected_source_semantic_hash=self.source_semantic_hash,
        )
        for entity_id, values in decoded_state.items():
            runtime.entities[entity_id].state.clear()
            runtime.entities[entity_id].state.update(values)
        return runtime


def _fail(code: str, path: str, message: str) -> None:
    raise TevScriptError(code, f"{path}: {message}")


def _object(value: Any, path: str) -> dict[str, Any]:
    if not isinstance(value, dict) or not all(isinstance(key, str) for key in value):
        _fail("TEVS_CHECKPOINT_V2_SHAPE", path, "expected object")
    return value


def _array(value: Any, path: str) -> list[Any]:
    if not isinstance(value, list):
        _fail("TEVS_CHECKPOINT_V2_SHAPE", path, "expected array")
    return value


def _exact_keys(value: Mapping[str, Any], path: str, expected: set[str]) -> None:
    observed = set(value)
    if observed != expected:
        _fail("TEVS_CHECKPOINT_V2_SHAPE", path, f"field set mismatch; missing={sorted(expected-observed)}, extra={sorted(observed-expected)}")


def _string(value: Any, path: str) -> str:
    if not isinstance(value, str):
        _fail("TEVS_CHECKPOINT_V2_SHAPE", path, "expected string")
    return value


def _local(value: Any, path: str) -> str:
    result = _string(value, path)
    if _LOCAL.fullmatch(result) is None:
        _fail("TEVS_CHECKPOINT_V2_IDENTIFIER", path, f"expected identifier, got {result!r}")
    return result


def _sha(value: Any, path: str) -> str:
    result = _string(value, path)
    if _HASH.fullmatch(result) is None:
        _fail("TEVS_CHECKPOINT_V2_SHA256", path, "expected lowercase SHA-256")
    return result
